### PR TITLE
CI: initial setup for IO Demo testing

### PR DESCRIPTION
**Pull request by amaslenn**
## What
Initial implementation of CI for `io_demo`

1. I removed all other checks to speed up testing and do not load regular CI.
2. This PR is to start the discussion about the implementation approach and test cases.
3. The review is required for `buildlib/io-demo.yml` only.

### CODE REVIEW

**Comment by alinask** [comment]
@amaslenn is there an example of a run?

**Comment by amaslenn** [comment]
> @amaslenn is there an example of a run?

Sure, see Azure CI result.

**Comment by alinask** [comment]
@amaslenn as discussed offline, please:
1. unite the code to two sections - one for server and one for client, so that it would be easier to reproduce what was run when we need to debug.
2. add an option to run io_demo with different arguments in separate iterations.
3. check if both client and server would be killed in case one's link goes down.

File: buildlib/az-io_demo.sh [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+source "${workspace}/az-helpers.sh"
+
+server_ip=${server_ip:=""}
+duration=${duration:=2}
+iface=${iface:="bond0"}
+
+## run server
+if [ "x$server_ip" = "x" ]; then
+    ip addr show ${iface}
+    server_ip=$(ip addr show ${iface} | awk '/inet / {print $2}' | awk -F/ '{print $1}')
```
Comment (yosefe) [inline buildlib/az-io_demo.sh:30]: add helper script to get ip

File: buildlib/az-io_demo.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    azure_set_variable "server_ip" "$server_ip"
+
+    source "${workspace}/../test/apps/iodemo/env"
+    if ! "${workspace}/../test/apps/iodemo/io_demo" >server.log 2>&1 & then
```
Comment (yosefe) [inline buildlib/az-io_demo.sh:33]: `>& server.log`

File: buildlib/az-io_demo.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        error "Failed to start server"
+    fi
+    server_pid=$!
+    echo "Server is running, PID=$server_pid"
+    azure_set_variable "server_pid" "$server_pid"
+
+    # double check the process is running
+    sleep 3
```
Comment (yosefe) [inline buildlib/az-io_demo.sh:41]: sleep 5

File: buildlib/io-demo.yml [hunk 4]
```diff
@@ -57,6 +57,13 @@ jobs:
     steps:
       - checkout: none
 
+      - bash: |
+          set -eEx
+          source ./buildlib/az-helpers.sh
+          ./buildlib/az-network-corrupter.sh initial_delay=1 cycles=2 downtime=2 uptime=5 >corrupter.log 2>&1
```
[the remark this answers is not in the document]
Comment (amaslenn) [reply buildlib/io-demo.yml:63]: 1. Done
2. These parameters should come from the test definition, I've modified the pipeline.

File: buildlib/az-io_demo.sh (hunk 1 above)
Comment (amaslenn) [reply buildlib/az-io_demo.sh:30]: Done

File: buildlib/az-io_demo.sh (hunk 2 above)
Comment (amaslenn) [reply buildlib/az-io_demo.sh:33]: Done

File: buildlib/az-io_demo.sh (hunk 3 above)
Comment (amaslenn) [reply buildlib/az-io_demo.sh:41]: Done

File: buildlib/az-network-corrupter.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+initial_delay=${initial_delay:=10}
+cycles=${cycles:=2}
+downtime=${downtime:=5}
+uptime=${uptime:=20}
+
+echo "Initial delay ${initial_delay} sec"
+sleep ${initial_delay}
+
+for i in $(seq 1 ${cycles} 1); do
+    echo "#$i Put it down! And sleep ${downtime}"; sleep "$downtime"
```
[the remark this answers is not in the document]
Comment (amaslenn) [reply buildlib/az-network-corrupter.sh:13]: Azure provides timestamp's toggle in its interface. Do we really need our own? Which format then?

File: buildlib/io-demo.yml [hunk 6]
```diff
@@ -100,3 +107,15 @@ jobs:
           fi
         displayName: Kill the server
         condition: always()
+
+      - bash: |
+          set -eEx
+          pid=$(corrupter_pid)
+
+          echo "Stopping  corrupter, PID=${pid}"
```
[the remark this answers is not in the document]
Comment (amaslenn) [reply buildlib/io-demo.yml:103]: Fixed.

File: buildlib/az-network-corrupter.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply buildlib/az-network-corrupter.sh:13]: would it provide timestamp for a script which runs in the background?
can be same FMT as Azure

File: buildlib/io-demo.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply buildlib/io-demo.yml]: 2. ok

File: buildlib/az-network-corrupter.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply buildlib/az-network-corrupter.sh:13]: same needed for server BTW

File: buildlib/io-demo.yml [hunk 7]
```diff
@@ -74,14 +74,14 @@ jobs:
             cycles=$(cycles) \
             downtime=$(downtime) \
             uptime=$(uptime) \
-            | add_timetamp >& corrupter.log
+            2>&1 | add_timestamp >& corrupter.log &
```
Comment (yosefe) [inline buildlib/io-demo.yml:77]: should be like this:
`|& add_timestamp >corrupter.log &`

File: buildlib/az-helpers.sh [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+# Report an issue to Azure pipeline and stop step execution
+function azure_log_issue() {
+    test "x$RUNNING_IN_AZURE" = "xno" && return
+    msg=$1
+    echo "##vso[task.logissue type=error]${msg}"
+    echo "##vso[task.complete result=Failed;]"
+}
+
+# Get IPv4 address of an interface
+function get_ip() {
```
Comment (alinask) [inline buildlib/az-helpers.sh:40]: why only ipv4? can we test ipv6 as well?

Comment (amaslenn) [reply buildlib/az-helpers.sh:40]: We surely can, but that would be different cases. When we add them this function should be updated or a new one added.

Comment (alinask) [reply buildlib/az-helpers.sh:40]: ok so let's keep it in a TODO list.
we had cases where there were bugs with ipv6 so it's important we add it please.

File: buildlib/io-demo.yml [hunk 9]
```diff
@@ -67,15 +67,17 @@ jobs:
...
           source ./buildlib/az-helpers.sh
           ./buildlib/az-network-corrupter.sh \
             initial_delay=$(initial_delay) \
             cycles=$(cycles) \
             downtime=$(downtime) \
             uptime=$(uptime) \
             |& add_timestamp >corrupter.log &
-          azure_set_variable "corrupter_pid" $$
+
+          corrupter_pid=$(ps -ef | grep -P '[n]etwork-corrupter' | awk '{print $2}')
```
Comment (yosefe) [inline buildlib/io-demo.yml:79]: use pgrep and pass user id as well

File: buildlib/io-demo.yml [hunk 10]
```diff
@@ -76,7 +76,7 @@ jobs:
             uptime=$(uptime) \
             |& add_timestamp >corrupter.log &
 
-          corrupter_pid=$(ps -ef | grep -P '[n]etwork-corrupter' | awk '{print $2}')
+          corrupter_pid=$(pgrep -u $USER 'network-corrupter')
```
Comment (yosefe) [inline buildlib/io-demo.yml:79]: "$USER"

File: buildlib/az-network-corrupter.sh [hunk 11]
```diff
@@ -2,14 +2,26 @@
...
 initial_delay=${initial_delay:=10}
-cycles=${cycles:=2}
+cycles=${cycles:=1000}
 downtime=${downtime:=5}
 uptime=${uptime:=20}
+reset=${reset:="no"}
+
+
+if [ "x$reset" = "xyes" ]; then
+    sudo /hpc/noarch/git_projects/swx_infrastructure/clusters/bin/manage_host_ports.sh "$(hostname)" "bond-up"
```
Comment (yosefe) [inline buildlib/az-network-corrupter.sh:12]: maybe use helper var for script path?

File: buildlib/az-io_demo.sh [hunk 12]
```diff
@@ -33,7 +34,7 @@ if [ "x$server_ip" = "x" ]; then
     if ! "${workspace}/../test/apps/iodemo/io_demo" |& add_timestamp >server.log & then
         error "Failed to start server"
     fi
-    server_pid=$(pgrep -u "$USER" io_demo | awk '{print $2}')
+    server_pid=$(pgrep -u "$USER" -f iodemo)
```
Comment (yosefe) [inline buildlib/az-io_demo.sh:37]: why need '-f' ? i think it could find unrelated process

Comment (amaslenn) [reply buildlib/az-io_demo.sh:37]: Couldn't find without `-f`, also had to use dir name, not a binary to see only one process and not parenting bash script.

Comment (yosefe) [reply buildlib/az-io_demo.sh:37]: can we make a more specific pattern?

File: buildlib/az-io_demo.sh [hunk 13]
```diff
@@ -34,7 +34,7 @@ if [ "x$server_ip" = "x" ]; then
     if ! "${workspace}/../test/apps/iodemo/io_demo" |& add_timestamp >server.log & then
         error "Failed to start server"
     fi
-    server_pid=$(pgrep -u "$USER" -f iodemo)
+    server_pid=$(pgrep -u $USER -f 'io_demo$')
```
Comment (yosefe) [inline buildlib/az-io_demo.sh:37]: same with corrupter?

Comment (amaslenn) [reply buildlib/az-io_demo.sh:37]: For corrupter it looks quite specific to me:
```sh
corrupter_pid=$(pgrep -u "$USER" -f 'network-corrupter')
```

And it doesn't spawn any new processes with similar names.

Comment (yosefe) [reply buildlib/az-io_demo.sh:37]: maybe 
`corrupter_pid=$(pgrep -u "$USER" -f 'network-corrupter$')`
or
`corrupter_pid=$(pgrep -u "$USER" -f 'network-corrupter.sh$')`

Comment (amaslenn) [reply buildlib/az-io_demo.sh:37]: Done.

Comment (amaslenn) [reply buildlib/az-io_demo.sh]: @yosefe had to revert both changes, `$` doesn't work when running in Azure, I don't know why.

**Comment by amaslenn** [comment]
@dmitrygx looks like `io_demo` resets the timer (`-l` option) on reconnects. This leads to undefined test time and in some cases test takes >2m and Azure stops it on timeout.

**Comment by dmitrygx** [comment]
> @dmitrygx looks like `io_demo` resets the timer (`-l` option) on reconnects. This leads to undefined test time and in some cases test takes >2m and Azure stops it on timeout.

@amaslenn yes, thank you for reporting this issue, it has to be fixed. will do

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by amaslenn** [comment]
@yosefe @dmitrygx let's make & merge the fix for `io_demo` before merging this PR, otherwise, CI might become very unstable.

**Comment by dmitrygx** [comment]
> @yosefe @dmitrygx let's make & merge the fix for `io_demo` before merging this PR, otherwise, CI might become very unstable.

@amaslenn make sense to me
will prepare PR today

**Comment by dmitrygx** [comment]
> > @yosefe @dmitrygx let's make & merge the fix for `io_demo` before merging this PR, otherwise, CI might become very unstable.
> 
> @amaslenn makes sense to me
> will prepare PR today

fix #5366 PRed

**Pull request merged**